[Decision needed] Added functions to compute DFT and IDFT matrices

oscargus opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
Simple but convenient function to compute the (I)DFT matrix.

#### Other comments
It will be interesting to see if the doctest passes as I have very limited idea how to format the output correctly for this.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* discrete
   * Added functions to compute DFT and IDFT matrices
<!-- END RELEASE NOTES -->

oscargus committed:
Added method to compute DFT and IDFT matrices

oscargus committed:
Corrected Doctest output format

sylee957 commented:
Dft matrix constructors may better be implemented in `MatrixSpecial` class in `matrices/common.py`, so that it can be possible to construct with different matrix classes easily.

oscargus commented:
@sylee957 I see your point. I'll have a look.

asmeurer commented:
CC @sidhantnagpal

oscargus commented:
I'd be happy to go with any implementation. I have some use cases where it is used on matrix form (to derive fast convolution algorithms), so to me that it is the main thing. It is clearly a bad idea to compute the result of a long DFT this way...

Should I move it to `MatrixSpecial` then? Using `entry`? Or will you @sidhantnagpal add your version?

sidhantnagpal commented:
Feel free to implement it. The performance can be improved incrementally. 
I should be able to have a look once it is implemented.

> Should I move it to `MatrixSpecial` then? Using `entry`?

I would wait a bit to see more thoughts on this.

oscargus committed:
Added dft_matix in MatrixSpecial

oscargus commented:
I added a version in `MatrixSpecial` now using `entry`. Faster than the previous approach, but probably not as fast as the approach of @sidhantnagpal .

I also changed it to have an optional parameter `inverse` instead of a separate `idft_matrix`method.

Is there a way to reuse the documentation of the two `dft_matrix` methods in `common.py` and `dense.py`?

oscargus commented:
The documentation in `dense` has a small error in one example. I will push the correct version when it is decided which approach should be kept.

This also indicates that doctest is not running for it as it passed on my computer before pushing. (It also indicates the benefit of having a single source of the documention...)

oscargus committed:
Corrected documention

oscargus committed:
Removed transform version as the doctests got confused

oscargus committed:
Final cleanup of transform approach

oscargus commented:
There are still some things to clean up. I have a local commit, but will not push until later to avoid running too many tests...

sylee957 commented:
I think the implementation on this does not match `fft` in discrete module because of the difference in sign.

However, I think this implementation is more correct than the implementation, and is consistent with numpy, so if it is kept, I think `fft` should be fixed.
I've addressed an issue at #15804

oscargus commented on sympy/matrices/dense.py:
```diff
@@ -1360,6 +1416,23 @@ def ones(*args, **kwargs):
 
+def dft_matrix(*args, **kwargs):
+    """Returns a matrix of ones with ``rows`` rows and ``cols`` columns;
+    if ``cols`` is omitted a square matrix will be returned.
```
This was an old copy of the method that I forgot to remove...

oscargus commented on sympy/matrices/dense.py:
```diff
@@ -1360,6 +1416,23 @@ def ones(*args, **kwargs):
+
+    zeros
+    eye
+    diag
```
I'll sort it out in the other methods (which I copied as a template).

oscargus committed:
Fixed review comments

sidhantnagpal commented on sympy/matrices/common.py:
```diff
@@ -653,6 +666,73 @@ def entry(i, j):
         return cls._new(rows, cols, entry)
 
+    @classmethod
+    def dft_matrix(kls, points, inverse=False, unitary=False, **kwargs):
```
On second thoughts, I think it maybe fine, as the name of the method name should convey the context.

> 2-D DFT

Most libraries refer to the multidimensional Fourier transform like fft2d/fft2 (or fftn for the n-dimensional case).

oscargus commented:
Thanks @sidhantnagpal for a thorough review and sorry for the newbie mistakes, but I learned things for the future for sure. I have pushed a new version now.

Regarding the documentation, I simply removed the one in `matrices/common.py`. I tried the `__doc__` approach earlier but didn't get it to work (recursive definitions somehow, do not remember which order I defined it). Seems to be better to have one good source of documentation somehow.

I hope I got everything in it.

sylee957 commented:
In my opinion, the loose functions in `matrices/dense.py` module are the remnants of some old designs, and it may not be a good moment to place some new functions inside.

For example, `ones`, `zeros`, `eye`, `diag`, `jordan_cell`, `multiply_elementwise`, and `GramSchmidt` are just 
 duplicate functions.
And also `carsotian`, `wronskian`, `hessian` may have to be placed in the similar context as `Matrix.jacobian`.

Once the special matrix constructors are set as classmethods, it is possible to use them without creating an instance, and users can take abbreviations like `dft = Matrix.dft_matrix` or whatever if they need, without much effort.

oscargus committed:
Merge branch 'master' into dftmatrix

oscargus commented:
I just found sympy.matrices.expressions.fourier which contains DFT and IDFT. Hence, this is not required. There is still a bit of potential to improve those, so I spend my efforts there.
